rskj-core/src/main/java/org/ethereum/core/BlockHeaderV1.java

```diff
@@ -0,0 +1,57 @@
+package org.ethereum.core;
+
+import co.rsk.core.BlockDifficulty;
+import co.rsk.core.Coin;
+import co.rsk.core.RskAddress;
+import org.ethereum.util.RLP;
+
+public class BlockHeaderV1 extends BlockHeader {
+    @Override
+    public byte getVersion() { return 0x1; }
+
+    private BlockHeaderExtensionV1 extension;
+    public BlockHeaderExtensionV1 getExtension() { return this.extension; }
+    public void setExtension(BlockHeaderExtension extension) { this.extension = (BlockHeaderExtensionV1) extension; }
+
+    public BlockHeaderV1(byte[] parentHash, byte[] unclesHash, RskAddress coinbase, byte[] stateRoot,
+                         byte[] txTrieRoot, byte[] receiptTrieRoot, byte[] logsBloom, BlockDifficulty difficulty,
+                         long number, byte[] gasLimit, long gasUsed, long timestamp, byte[] extraData,
+                         Coin paidFees, byte[] bitcoinMergedMiningHeader, byte[] bitcoinMergedMiningMerkleProof,
+                         byte[] bitcoinMergedMiningCoinbaseTransaction, byte[] mergedMiningForkDetectionData,
+                         Coin minimumGasPrice, int uncleCount, boolean sealed,
+                         boolean useRskip92Encoding, boolean includeForkDetectionData, byte[] ummRoot) {
+        super(parentHash,unclesHash, coinbase, stateRoot,
+                txTrieRoot, receiptTrieRoot, difficulty,
+                number, gasLimit, gasUsed, timestamp, extraData,
+                paidFees, bitcoinMergedMiningHeader, bitcoinMergedMiningMerkleProof,
+                bitcoinMergedMiningCoinbaseTransaction, mergedMiningForkDetectionData,
+                minimumGasPrice, uncleCount, sealed,
+                useRskip92Encoding, includeForkDetectionData, ummRoot);
+
+        this.extension = new BlockHeaderExtensionV1(logsBloom);
+    }
+
+    @Override
+    public byte[] getLogsBloom() {
+        return this.extension.getLogsBloom();
+    }
+
+    public void setLogsBloom(byte[] logsBloom) {
+        /* A sealed block header is immutable, cannot be changed */
+        if (this.sealed) {
+            throw new SealedBlockHeaderException("trying to alter logs bloom");
+        }
+        this.hash = null;
+
+        this.extension.setLogsBloom(logsBloom);
+    }
+
+    @Override
+    public byte[] getLogsBloomFieldEncoded() {
+        byte[] ecnoded = new byte[Bloom.BLOOM_BYTES];
+        ecnoded[0] = 0x1;
```

Comment:
What is this 0x1 at position 0 ?

Reply:
Why is ecnoded of size BLOOM_BYTES (256) when it should be 33 (if you add the 0x1)

Reply:
Why is the field called "ecnoded" and not "encoded" ?

Reply:
Based on the RSKIP compressed encoding description https://github.com/rsksmart/RSKIPs/blob/miniheader/IPs/RSKIP351.md#compressed-header

> When serializing the block header version 1 to obtain the block hash, the field logsBloom is replaced by an RLP list of
>  - The blockVersion field (integer) (1 byte)
>  - The Keccak hash of the bloom filter data and the txExecutionSublistsEdges as a list of two elements (32 bytes).

Additionally, based on RSKIP 194

> To avoid forcing a costly firmware upgrade for all PowHSMs, the blockVersion is stored in the logsBloom filter, which is unchecked by the hardware.

The “get logs bloom field” will return 256 bytes that go in the rlp encoding of the compressed header. That is why it is created with the logs bloom size. It was easier to modify fixed size bytes than creating a new rlp list that fits exactly in a given size

I underdtand we need to keep the same 256 bytes size for that field. If not, we can just remove the padding and keep it 33 bytes size

(ecnoded is a typo)

Reply:
I log bloom field does not need to be 256 bytes in length. In fact, we aim to reduce it because nobody uses it.
Second, I don't think the version should be stored inside the logBloom field: it think it deserves a separate field in the RLP. That gives much more flexibility (i.e. more than one byte for versioning)

Reply:
Let me check again the RSKIP

Reply:
Ok so required changes are:
- make logs bloom field 32 bytes size
- add new rlp field for version

(it might be easier to implement logs bloom field as 32 bytes for the hash + rest of bytes for version though, but i like more your proposal)